.travis.yml

```diff
@@ -63,34 +69,55 @@ matrix:
   fast_finish: true
 
 before_install:
+  # Setup the cache directory paths
+  - if [[ ! -d test-deps ]]; then mkdir test-deps ; fi
+  - export PYTHON_INSTALL_DIR=`pwd`/test-deps/python
+  - export GNUTLS_INSTALL_DIR=`pwd`/test-deps/gnutls
+  - export PRLIMIT_INSTALL_DIR=`pwd`/test-deps/prlimit
+  - export SAW_INSTALL_DIR=`pwd`/test-deps/saw
+  - export Z3_INSTALL_DIR=`pwd`/test-deps/z3
+  - export OPENSSL_1_1_0_INSTALL_DIR=`pwd`/test-deps/openssl-1.1.0
   # Install GCC 6 if on OSX
   - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then   .travis/install_osx_dependencies.sh ; fi
   # Set GCC 6 as Default on both Ubuntu and OSX
   - alias gcc=$(which gcc-6)
   # Install latest version of clang, clang++, and llvm-symbolizer and add them to beginning of PATH. Needed for fuzzing.
   - if [[ "$LATEST_CLANG" == "true" ]]; then (.travis/install_clang.sh `pwd`/clang-download `pwd`/clang-latest $TRAVIS_OS_NAME) && export PATH=`pwd`/clang-latest/bin:$PATH ; fi
-  # Install SAW with Z3 on Linux only
-  - if [[ "$SAW" == "true" ]]; then .travis/install_saw.sh ; fi
-  - export PATH=$PATH:$PWD/saw-0.2-2016-11-02-Ubuntu14.04-64/bin:$PWD/z3
 
+# Install missing test dependencies. If the install directory already exists, cached artifacts will be used
+# for that dependency.
 install:
   # Download and Install LibFuzzer
   - if [[ "$TESTS" == "fuzz" ]]; then .travis/install_libFuzzer.sh `pwd`/fuzz_dependencies/libFuzzer-download `pwd`/fuzz_dependencies $TRAVIS_OS_NAME ; fi
-  # Download and Install Openssl
-  - .travis/install_openssl_1_1_0.sh `pwd`/libcrypto-build `pwd`/libcrypto-root $TRAVIS_OS_NAME > /dev/null
-  # Install python linked with our compiled Openssl for integration tests
-  - if [[ "$TESTS" == "integration" ]]; then mkdir python-build python && .travis/install_python.sh `pwd`/libcrypto-root `pwd`/python-build `pwd`/python > /dev/null && export PATH=`pwd`/python/bin:$PATH ; fi
+  # Download and Install Openssl 1.1.0
+  - if [[ ! -d "$OPENSSL_1_1_0_INSTALL_DIR" ]]; then .travis/install_openssl_1_1_0.sh /tmp $OPENSSL_1_1_0_INSTALL_DIR $TRAVIS_OS_NAME > /dev/null ; fi
+  # Install python linked with the latest Openssl for integration tests
+  - if [[ ! -d "$PYTHON_INSTALL_DIR" ]] && [[ "$TESTS" == "integration" ]]; then mkdir -p $PYTHON_INSTALL_DIR && .travis/install_python.sh $OPENSSL_1_1_0_INSTALL_DIR /tmp $PYTHON_INSTALL_DIR > /dev/null ; fi
```

Comment:
If we are caching our dependencies now, do we still need the logic to only install certain dependencies in certain cases?

It would make dependency management a lot simpler to just install everything, pay the price once, then have everything cached rather than having a different "if" statement with a different condition on every line. It would also make https://github.com/awslabs/s2n/issues/385 a lot easier if we just had one script that installed all Ubuntu dependencies.